lib/segment/src/index/field_index/field_index_base.rs

```diff
@@ -73,42 +72,22 @@ pub trait ValueIndexer<T> {
     }
 
     /// Add point with payload to index
-    fn add_point(
-        &mut self,
-        id: PointOffsetType,
-        payload: &MultiValue<&Value>,
-    ) -> OperationResult<()> {
+    fn add_point(&mut self, id: PointOffsetType, payload: &[&Value]) -> OperationResult<()> {
         self.remove_point(id)?;
-        match payload {
-            MultiValue::Multiple(values) => {
-                let mut flatten_values: Vec<_> = vec![];
-
-                for value in values {
-                    match value {
-                        Value::Array(values) => {
-                            flatten_values.extend(values.iter().flat_map(|x| self.get_value(x)));
-                        }
-                        _ => {
-                            if let Some(x) = self.get_value(value) {
-                                flatten_values.push(x);
-                            }
-                        }
-                    }
+        let mut flatten_values: Vec<_> = vec![];
+        for value in payload.iter() {
+            match value {
+                Value::Array(values) => {
+                    flatten_values.extend(values.iter().flat_map(|x| self.get_value(x)));
                 }
-                self.add_many(id, flatten_values)
-            }
-            MultiValue::Single(Some(Value::Array(values))) => {
-                self.add_many(id, values.iter().flat_map(|x| self.get_value(x)).collect())
-            }
-            MultiValue::Single(Some(value)) => {
-                if let Some(x) = self.get_value(value) {
-                    self.add_many(id, vec![x])
-                } else {
-                    Ok(())
+                _ => {
+                    if let Some(x) = self.get_value(value) {
+                        flatten_values.push(x);
+                    }
                 }
             }
-            MultiValue::Single(None) => Ok(()),
         }
+        self.add_many(id, flatten_values)
     }
```

Comment:
It's no worse than the previous implementation. Breakdown of cases:

Old implementation:
- `Multiple(empty vec)`: no allocations
- `Multiple(_)`: `flatten_values` is allocated
- `Single(Some(array))`: allocation due to `.collect()` being used
- `Single(Some(_))`: allocation due to `vec![x]` being used
- `Single(None)`: no allocations

New implementation:
If `flatten_values` is empty, no allocation is made (since an empty vector doesn't allocate heap memory), otherwise the memory is allocated.

---

However, I think it's could be improved if we change the implementations of `add_many()` to accept a slice or an iterator instead of a vec. But I propose to do this outside of this PR.

Reply:
> However, I think it's could be improved if we change the implementations of add_many() to accept a slice or an iterator instead of a vec. But I propose to do this outside of this PR.

:+1: